### PR TITLE
[gcs][kv-regression][2/n] Add multi get for internal kv to reduce rpc request

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

Internal KV is moved from redis to gcs since we'd like remove redis dependence. But this introduced a performance regression when GCS load is very high.

Previously: Worker <-> Redis
And now: Worker <-> GCS <-> Redis

So internal traffic is competing with other traffic in GCS. In this PR we did several optimizations to mitigate this:
- Use multi-get to reduce traffic number to gcs when worker starts.
- Execute internal kv request in CQ threads since they are cheap to reduce latency
- Increase grpc server thread number to 4 so that it can make kv performance better since now they are consumed in CQ threads.

This is tried as well: use a dedicated thread for kv. but it's not working very well maybe we need a thread pool for that, so in the end, I think just increasing grpc server thread number is better. And in this PR, we are able to execute functions in GRPC directly.

The new result is:
```
  actors_per_second = 395.02975194777275
  num_actors = 10000
  time = 25.3145489692688
```

which is still slower than previous (time=20s). The reason is that now it's doing more things than before.

TODO: once redis is removed, we'll adopt a memory table and we'll be able to fix this. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Why not just making its own rpc server like this https://github.com/ray-project/ray/blob/da57c99f47b0ad2a7d8d82248a8b88cbc8c71d2d/src/ray/gcs/gcs_server/gcs_server.cc#L227

**Comment by ericl** [comment]
Can you add a description to this PR?

**Comment by iycheng** [comment]
> Can you add a description to this PR?

@ericl  it's still wip. It's not ready for a review now.

**Comment by iycheng** [comment]
> Quick Q. Why not just making its own rpc server like this? Do you see some disadvantages from this approach?
> 
> https://github.com/ray-project/ray/blob/da57c99f47b0ad2a7d8d82248a8b88cbc8c71d2d/src/ray/gcs/gcs_server/gcs_server.cc#L227

@rkooo567  I tried, it's not that good. Check [here](https://github.com/ray-project/ray/pull/20788/files/238410fb8e313006e1a6da953dfba56bafb457aa)

**Comment by rkooo567** [comment]
@iycheng I believe we can trivially implement thread pool with separate RPC server using the same approach as 

https://github.com/ray-project/ray/blob/4ef0d4a37a42c529af98b0cfb31e505b51088395/src/ray/object_manager/object_manager.cc#L148

I think this seems to be better than allowing to run function directly in grpc server. It makes the already complicated code path harder to understand and we cannot utilize the existing nice tooling around instrumented io threads

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc [hunk 1]
```diff
@@ -33,6 +33,25 @@ void GcsInternalKVManager::HandleInternalKVGet(
       }));
 }
 
+void GcsInternalKVManager::HandleInternalKVMultiGet(
+    const rpc::InternalKVMultiGetRequest &request, rpc::InternalKVMultiGetReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  auto cnt = std::make_shared<int>(request.keys_size());
+  for (int i = 0; i < request.keys_size(); ++i) {
+    std::vector<std::string> cmd = {"HGET", request.keys(i), "value"};
+    RAY_CHECK_OK(redis_client_->GetPrimaryContext()->RunArgvAsync(
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_kv_manager.cc:42]: Is this thread safe? Especially `redis_client_->GetPrimaryContext()->RunArgvAsync(`?

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc [hunk 2]
```diff
@@ -33,6 +33,25 @@ void GcsInternalKVManager::HandleInternalKVGet(
...
+        cmd, [i, reply, send_reply_callback, cnt](auto redis_reply) {
+          if (!redis_reply->IsNil()) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_kv_manager.cc:44]: Is this callback running in the main io service thread?

File: src/ray/protobuf/gcs_service.proto [hunk 3]
```diff
@@ -557,6 +566,7 @@ message InternalKVKeysReply {
 // Service for KV storage
 service InternalKVGcsService {
   rpc InternalKVGet(InternalKVGetRequest) returns (InternalKVGetReply);
+  rpc InternalKVMultiGet(InternalKVMultiGetRequest) returns (InternalKVMultiGetReply);
```
Comment (rkooo567) [inline src/ray/protobuf/gcs_service.proto:569]: Can you add docstring to explain how it is different from `InternalKVGet `? Especially if we fail to obtain one of keys, what's going to happen?

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc (hunk 1 above)
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_kv_manager.cc:42]: it is

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_kv_manager.cc:44]: yes and that's bad due to our model

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_kv_manager.cc:44]: The callback inside doesn't touch any private field right? Can you add comment to warn not to do that?

Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_kv_manager.cc:44]: sure. btw, i think in either way it won't be fixed easily (in place running or having a separate thread running this)

**Comment by iycheng** [comment]
> @iycheng I believe we can trivially implement thread pool with separate RPC server using the same approach as
> 
> https://github.com/ray-project/ray/blob/4ef0d4a37a42c529af98b0cfb31e505b51088395/src/ray/object_manager/object_manager.cc#L148
> 
> I think this seems to be better than allowing to run function directly in grpc server. It makes the already complicated code path harder to understand and we cannot utilize the existing tooling around instrumented io service.

@rkooo567 Let's check the pros and cons here:

This PR:
- request -> CQ -> Redis
  - pros: no extra things about thread pool, things are just implemented. it can also make CQ faster in case protobuf ser/de are expensive (most of the time it's ok).
  - cons: now threads are running with different logic. previously it's running with poll only, but now it also has business logic there which make it harder to maintain.

With your proposal:
- request -> cq -> thread pool -> redis
  - pros: it's cleaner in architecture, each thread only works with the same things.
  - cons: most of the time threads are just idle there and it can't help cq as well. 

I think a little bit about this, feeling what you proposed might be better, but I also feel for pattern 1) sometimes it's also useful especially when the ops are super cheap.

I'm going to change threading model after gcs ha kv work, so I'll go with your proposal first.

**Comment by ericl** [comment]
+1 on Sang's suggestion, since it eliminates contention between the main loop and kv ops. With this change, we can probably avoid increasing the number of threads as well, right?

File: python/ray/tests/test_basic_3.py [hunk 4]
```diff
@@ -433,5 +433,14 @@ def run():
     ray.get(verify_job_id.remote(job_id, True))
 
 
+def test_kv_multi_get(ray_start_regular):
```
Comment (ericl) [inline python/ray/tests/test_basic_3.py:436]: Could we split the multi-get changes into a separate PR?

File: src/ray/common/ray_config_def.h [hunk 5]
```diff
@@ -248,7 +248,7 @@ RAY_CONFIG(int32_t, maximum_profile_table_rows_count, 10 * 1000)
 RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
 
 /// Number of threads used by rpc server in gcs server.
-RAY_CONFIG(uint32_t, gcs_server_rpc_server_thread_num, 1)
+RAY_CONFIG(uint32_t, gcs_server_rpc_server_thread_num, 4)
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:251]: Would be great to not have to change this (doesn't it break k8s?) Should be a separate PR if we do decide to do it.

Comment (iycheng) [reply src/ray/common/ray_config_def.h:251]: it doesn't break k8s. we fixed that in the test

File: python/ray/tests/test_basic_3.py (hunk 4 above)
Comment (iycheng) [reply python/ray/tests/test_basic_3.py:436]: If you wish, but I'd like to have a separate PR for other changes and convert this one for multi-get only.

**Comment by iycheng** [comment]
I'll comeback to this after https://github.com/ray-project/ray/pull/20922 merged

**Comment by iycheng** [comment]
Close this PR since it's not necessary. I'll reopen it if we need it.
